<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using pafr to read, manipulate and plot genome alignments in the PAF format • pafr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using pafr to read, manipulate and plot genome alignments in the PAF format">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pafr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction_to_pafr.html">Using pafr to read, manipulate and plot genome alignments in the PAF format</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dwinter/pafr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using pafr to read, manipulate and plot genome alignments in the PAF format</h1>
                        <h4 class="author">David Winter</h4>
            
            <h4 class="date">2019-10-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dwinter/pafr/blob/master/vignettes/Introduction_to_pafr.Rmd"><code>vignettes/Introduction_to_pafr.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction_to_pafr.Rmd</code></div>

    </div>

    
    
<div id="reading-data-with-read_paf" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-data-with-read_paf" class="anchor"></a>Reading data with <code>read_paf</code>
</h2>
<p><code>pafr</code> deals with data stored in the Pairwise mApping Format (PAF), as used by <code>minimap2</code> and other whole-genome aligners. PAF is a plain text tabular format where each row represents an alignment between two sequences. Each <code>.paf</code> file has at least the following 12 columns.</p>
<table class="table">
<thead><tr class="header">
<th><strong>Column</strong></th>
<th><strong>Data type</strong></th>
<th><strong>Description</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>string</td>
<td>Query sequence name</td>
</tr>
<tr class="even">
<td>2</td>
<td>int</td>
<td>Query sequence length</td>
</tr>
<tr class="odd">
<td>3</td>
<td>int</td>
<td>Query start coordinate (0-based)</td>
</tr>
<tr class="even">
<td>4</td>
<td>int</td>
<td>Query end coordinate (0-based)</td>
</tr>
<tr class="odd">
<td>5</td>
<td>char</td>
<td>‘+’ if query/target on the same strand; ‘-’ if opposite</td>
</tr>
<tr class="even">
<td>6</td>
<td>string</td>
<td>Target sequence name</td>
</tr>
<tr class="odd">
<td>7</td>
<td>int</td>
<td>Target sequence length</td>
</tr>
<tr class="even">
<td>8</td>
<td>int</td>
<td>Target start coordinate on the original strand</td>
</tr>
<tr class="odd">
<td>9</td>
<td>int</td>
<td>Target end coordinate on the original strand</td>
</tr>
<tr class="even">
<td>10</td>
<td>int</td>
<td>Number of matching bases in the mapping</td>
</tr>
<tr class="odd">
<td>11</td>
<td>int</td>
<td>Number bases, including gaps, in the mapping</td>
</tr>
<tr class="even">
<td>12</td>
<td>int</td>
<td>Mapping quality (0-255 with 255 for missing)</td>
</tr>
</tbody>
</table>
<p>In addition, each row can have a variable number of ‘tags’, which share a simliar format to those used SAM files <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">See the SAM specification (pdf)</a>.</p>
<p><code>pafr</code> provides the function <code>read_paf</code> to read alignments into an R session.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(pafr)</code></pre></div>
<pre><code>## Loading required package: ggplot2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_fungal_alignment &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"fungi.paf"</span>, <span class="dt">package =</span> <span class="st">"pafr"</span>)
ali &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_paf.html">read_paf</a></span>(path_to_fungal_alignment)</code></pre></div>
<p>By default, the object returned by read_paf behaves almost exactly like a base R <code>data.frame</code>. In fact, the only difference is this object prints a nice summary of the object content when called directly or printed (rather than the thousands of lines it might contain):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ali</code></pre></div>
<pre><code>## pafr object with 2501 alignments (36.5Mb)
##  8 query seqs
##  8 target seqs
##  11 tags: NM, ms, AS, nn, tp, cm, s1, s2, dv, cg, zd</code></pre>
<p>In this case there is a total of 35Mb of aligned-genome, coming from 8 query and 8 target sequences.<code>read_paf</code> gives the 12 core columns of the PAF format the following column names.</p>
<table class="table">
<thead><tr class="header">
<th>Column</th>
<th>Name</th>
<th>Data type</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>qname</td>
<td>string</td>
<td>Query sequence name</td>
</tr>
<tr class="even">
<td>2</td>
<td>qlen</td>
<td>int</td>
<td>Query sequence length</td>
</tr>
<tr class="odd">
<td>3</td>
<td>qstart</td>
<td>int</td>
<td>Query start coordinate (0-based)</td>
</tr>
<tr class="even">
<td>4</td>
<td>qend</td>
<td>int</td>
<td>Query end coordinate (0-based)</td>
</tr>
<tr class="odd">
<td>5</td>
<td>strand</td>
<td>char</td>
<td>‘+’ if query/target on the same strand; ‘-’ if opposite</td>
</tr>
<tr class="even">
<td>6</td>
<td>tname</td>
<td>string</td>
<td>Target sequence name</td>
</tr>
<tr class="odd">
<td>7</td>
<td>tlen</td>
<td>int</td>
<td>Target sequence length</td>
</tr>
<tr class="even">
<td>8</td>
<td>tstart</td>
<td>int</td>
<td>Target start coordinate on the original strand</td>
</tr>
<tr class="odd">
<td>9</td>
<td>tend</td>
<td>int</td>
<td>Target end coordinate on the original strand</td>
</tr>
<tr class="even">
<td>10</td>
<td>nmatch</td>
<td>int</td>
<td>Number of matching bases in the mapping</td>
</tr>
<tr class="odd">
<td>11</td>
<td>alen</td>
<td>int</td>
<td>Number bases, including gaps, in the mapping</td>
</tr>
<tr class="even">
<td>12</td>
<td>mapq</td>
<td>int</td>
<td>Mapping quality (0-255 with 255 for missing)</td>
</tr>
</tbody>
</table>
<p>In addition, any tags present in the file will be included as a column. The tags used by <code>minimap2</code> <a href="https://lh3.github.io/minimap2/minimap2.html">are documented here</a> or in the man page for that program. One interesting tag for this alignment is <code>dv</code>, which encodes the approximate per-base difference between query and target sequence. We can use this tag to visualize the relatinship between alignment-length and sequence divergence. Because the alignment data is stored in a <code>data.frame</code> we can pass it directly to <code>ggplot2</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggpubr)</code></pre></div>
<pre><code>## Loading required package: magrittr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(ali, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(alen, dv)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha=</span><span class="fl">0.6</span>, <span class="dt">colour=</span><span class="st">"steelblue"</span>, <span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="st">"Alignment length (kb)"</span>, <span class="dt">label =</span>  <span class="cf">function</span>(x) x<span class="op">/</span><span class="st"> </span><span class="fl">1e3</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span>(<span class="st">"Per base divergence"</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="https://rdrr.io/pkg/ggpubr/man/theme_pubr.html">theme_pubr</a></span>()</code></pre></div>
<pre><code>## Warning: Removed 52 rows containing missing values (geom_point).</code></pre>
<p><img src="Introduction_to_pafr_files/figure-html/alen-1.png" width="576" style="display: block; margin: auto;"></p>
<p>There are a lot of short alignments, some of which are very divergent. But there are also so very long alignments, which all show high similarity.</p>
<p>Because the <code>pafr</code> object is effectively a <code>data.frame</code>, we can use normal R functions to inspect or analyse it. For example, let’s calculate the mean divergence-level for alignments featuring each query sequence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by_q &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span>(dv <span class="op">~</span><span class="st"> </span>qname, <span class="dt">data=</span>ali, <span class="dt">FUN=</span>mean)
knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(by_q)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">qname</th>
<th align="right">dv</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Q_chr1</td>
<td align="right">0.0514848</td>
</tr>
<tr class="even">
<td align="left">Q_chr2</td>
<td align="right">0.0493231</td>
</tr>
<tr class="odd">
<td align="left">Q_chr3</td>
<td align="right">0.0556407</td>
</tr>
<tr class="even">
<td align="left">Q_chr4</td>
<td align="right">0.0543924</td>
</tr>
<tr class="odd">
<td align="left">Q_chr5</td>
<td align="right">0.0501364</td>
</tr>
<tr class="even">
<td align="left">Q_chr6</td>
<td align="right">0.0560656</td>
</tr>
<tr class="odd">
<td align="left">Q_chr7</td>
<td align="right">0.0479018</td>
</tr>
<tr class="even">
<td align="left">Q_chrm</td>
<td align="right">0.0185438</td>
</tr>
</tbody>
</table>
<p>Interestingly enough, <code>Q_chrm</code> is the mitochondrial genome, so appears the mitochrondial genome diplays less diveregnce than any of the chromosomes in this case.</p>
</div>
<div id="filtering-and-subsetting-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-and-subsetting-alignments" class="anchor"></a>Filtering and subsetting alignments</h2>
<p>Often the first thing you will want to do after reading in some data is get rid of low-quality, short and otherwise not-ideal alignments. <code>minimap2</code> uses the <code>tp</code> (type of alignment) tag to mark secondary alignments (i.e. possible alignments between the query and target sequence that are not the best possible alignment for those regions). The function <code>filter_secondary_alignments</code> removes non-primary alignments from a <code>pafr object</code>. Let’s use it to filter our <code>ali</code> object and see how many alignments (i.e. rows of data) we lose as a result:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">prim_alignment &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filter_secondary_alignments.html">filter_secondary_alignments</a></span>(ali)
<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(ali) <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(prim_alignment)</code></pre></div>
<pre><code>## [1] 339</code></pre>
<p>In some cases you might want to remove alignments with low mapping quality, high divergence, short length or some other property. Again, because the <code>pafr</code> object behaves like a <code>data.frame</code> you can use your favourite base <code>R</code> or <code>tidyverse</code> functions to do this. Here I remove short alignments and those with a low mapping quality score:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">long_ali &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(ali, alen <span class="op">&gt;</span><span class="st"> </span><span class="fl">1e4</span> <span class="op">&amp;</span><span class="st"> </span>mapq <span class="op">&gt;</span><span class="st"> </span><span class="dv">40</span>)
long_ali</code></pre></div>
<pre><code>## pafr object with 419 alignments (26.8Mb)
##  8 query seqs
##  8 target seqs
##  11 tags: NM, ms, AS, nn, tp, cm, s1, s2, dv, cg, zd</code></pre>
<p>Note, this has removed quite a lot of data (thousands of alignments and ~ 10Mb of sequence).</p>
</div>
<div id="plotting-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-alignments" class="anchor"></a>Plotting alignments</h2>
<p>One of the major aims of <code>pafr</code> is to generate compelling visualisatoins of genomic alignments. To this end, the he package implements three types of genomic visualizatoin, making use of the <code>ggplot2</code> plotting environment. Alhough <code>ggplot2</code> is powerful visualizatoin tool, we found producing high-quality plots of genome alignments was difficult, requring the use of many different layers and transformations to datasets. The visualisation functions provided by <code>pafr</code> automate the process of transoformating data and adding layers to a plot. Each of tehse functoins returns <code>ggplot2</code> plots, which can then be modified by the addition of ggplot layers, scales and themes.</p>
<div id="dotplot" class="section level3">
<h3 class="hasAnchor">
<a href="#dotplot" class="anchor"></a>Dotplot</h3>
<p>Often the most useful visualization of a genome alignment is the <a href="https://en.wikipedia.org/wiki/Dot_plot_(bioinformatics)">dotplot</a>. <code>pafr</code> makes it easy to produce one of these from a .paf alignment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/dotplot.html">dotplot</a></span>(prim_alignment)</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/dotplot_plain-1.png" width="576" style="display: block; margin: auto;"></p>
<div id="altering-the-order-of-sequences-in-the-dotplot" class="section level4">
<h4 class="hasAnchor">
<a href="#altering-the-order-of-sequences-in-the-dotplot" class="anchor"></a>Altering the order of sequences in the dotplot</h4>
<p>The defaul plot is quite bare, with each aligned segment shown as a dark line and the borders of the sequences in the query and target genomes are dashed lines. Additional arguments to <code>dotplot</code> allow us to modify the plot. In this case we can add labels for each query and target squence (<code>label_seqs</code>) and changes the order in which target sequences appear. Because the dotplot being produced here is a ggplot object, we can use <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw()</a></code> to change the plots theme too.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/dotplot.html">dotplot</a></span>(prim_alignment, <span class="dt">label_seqs=</span><span class="ot">TRUE</span>, <span class="dt">order_by=</span><span class="st">"qstart"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/dot_qstart-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The argument <code>order_by</code> takes three possible values: “size”, “qstart” or “provided”. “Size”(the default value) simply lines up query and target sequences from largest to smallest. “qstart” keeps the query sequences ordered by size, but rearranges the targets by where they match to query sequences. For example, <code>T_chr5</code> is moved to the second sequence in the above plot, as it matches to the first query chromosome. If <code>ordered_by</code> is set to “provided” you need provide a list with two elements (the order of query and then target sequences) to the function. This approach can also be used to down-sample an alignment to only a few focal sequences.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">to_keep &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
        <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Q_chr1"</span>, <span class="st">"Q_chr5"</span>, <span class="st">"Q_chr4"</span>, <span class="st">"Q_chr6"</span>),
        <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"T_chr2"</span>, <span class="st">"T_chr5"</span>, <span class="st">"T_chr3"</span>, <span class="st">"T_chr6"</span>)
)
<span class="kw"><a href="../reference/dotplot.html">dotplot</a></span>(prim_alignment, <span class="dt">label_seqs=</span><span class="ot">TRUE</span>, <span class="dt">order_by=</span><span class="st">"provided"</span>, <span class="dt">ordering=</span>to_keep)</code></pre></div>
<pre><code>## Warning: Dropping data from sequences absent from ordering:
##  Q_chr2,Q_chr3,Q_chr7,Q_chrm</code></pre>
<pre><code>## Warning: Dropping data from sequences absent from ordering:
##  T_chr1,T_chr7,T_chr4,T_chrm</code></pre>
<p><img src="Introduction_to_pafr_files/figure-html/dot_provided-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="adding-annotations" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-annotations" class="anchor"></a>Adding annotations</h4>
<p>You may have additional information about one or both of your genomes that you want to use to annotate your dotplots. In this case, we know the locations of the centromeres in the query genome. We can use <code>read_bed</code> to load these locations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path_to_centro &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"Q_centro.bed"</span>, <span class="dt">package =</span> <span class="st">"pafr"</span>)
centro &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(path_to_centro)
knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(centro))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">chrom</th>
<th align="right">start</th>
<th align="right">end</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Q_chr1</td>
<td align="right">1305119</td>
<td align="right">1325119</td>
</tr>
<tr class="even">
<td align="left">Q_chr2</td>
<td align="right">3311000</td>
<td align="right">3331000</td>
</tr>
<tr class="odd">
<td align="left">Q_chr5</td>
<td align="right">1780000</td>
<td align="right">1800000</td>
</tr>
<tr class="even">
<td align="left">Q_chr6</td>
<td align="right">1005337</td>
<td align="right">1025337</td>
</tr>
<tr class="odd">
<td align="left">Q_chr4</td>
<td align="right">951000</td>
<td align="right">971000</td>
</tr>
<tr class="even">
<td align="left">Q_chr7</td>
<td align="right">449000</td>
<td align="right">469000</td>
</tr>
</tbody>
</table>
<p>In <code>ggplot</code> style, we use the <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> symbol to add the results of <code>highlight_query</code> to a dotplot. Note, in order for the plot to work both the <code>ordered_by</code> and <code>ali</code> arguments must match between the call to <code>dotplot</code> and <code>highlight_query</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/dotplot.html">dotplot</a></span>(prim_alignment, <span class="st">"qstart"</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="../reference/highlight_dotplot.html">highlight_query</a></span>(<span class="dt">ali=</span>prim_alignment, <span class="dt">bed=</span>centro, <span class="dt">ordered_by=</span><span class="st">"qstart"</span>)</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/highlight-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It’s not apparent from the very small intervals plotted above, but ‘highlights’ here are a filled rectangle. Plotting a larger rectangle for the target genome makes it more clear.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">interval &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">chrom=</span><span class="st">"T_chr3"</span>, <span class="dt">start =</span> <span class="dv">2000000</span>, <span class="dt">end=</span> <span class="dv">3000000</span>)
<span class="kw"><a href="../reference/dotplot.html">dotplot</a></span>(prim_alignment, <span class="dt">label_seqs=</span><span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="../reference/highlight_dotplot.html">highlight_target</a></span>(prim_alignment, interval)</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/highlight_t-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="synteny-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#synteny-plot" class="anchor"></a>Synteny plot</h3>
<p>The dotplot gives us a ‘whole genome’ view of a genomic alignment. Very often we will also want to zoom down to look at how regions of two particular chromosomes are aligned to each other. The function <code>plot_synteny</code> provides a way to visualize alignments at this level. Lets compare regions of query chromosome 3 and target chromosome 4.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_synteny.html">plot_synteny</a></span>(long_ali, <span class="dt">q_chrom=</span><span class="st">"Q_chr3"</span>, <span class="dt">t_chrom=</span><span class="st">"T_chr4"</span>, <span class="dt">centre=</span><span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
<p>In this plot, each sequence is represented as a white box and each alignment is a grey segment connecting the regions of those sequences that align to each other. In this case, the alignment is quite straightforward. The plot looks less clear when most of the alignments are in reverse-and-complement.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_synteny.html">plot_synteny</a></span>(long_ali, <span class="dt">q_chrom=</span><span class="st">"Q_chr5"</span>, <span class="dt">t_chrom=</span><span class="st">"T_chr5"</span>, <span class="dt">centre=</span><span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-3-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The simple answer in these cases is to set the argument <code>rc</code> to <code>TRUE</code>, forcing the target sequence to be flipped.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_synteny.html">plot_synteny</a></span>(long_ali, <span class="dt">q_chrom=</span><span class="st">"Q_chr5"</span>, <span class="dt">t_chrom=</span><span class="st">"T_chr5"</span>, <span class="dt">centre=</span><span class="ot">TRUE</span>, <span class="dt">rc=</span><span class="ot">TRUE</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"> ### Coverage plot</p>
<p>The function <code>plot_coverage</code> gives a useful way to see how much of a given genome is included in a genome alignment. By default, it displays each sequence in the target chromosome as a rectangular box, with shaded regions represented parts of the target genome that are included in an alignment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_coverage.html">plot_coverage</a></span>(prim_alignment)   </code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_plot-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It is also possible to produce the same plot using the query genome as the reference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_coverage.html">plot_coverage</a></span>(prim_alignment, <span class="dt">target=</span><span class="ot">FALSE</span>)   </code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_target-1.png" width="576" style="display: block; margin: auto;"></p>
<div id="coverage-by-query-sequence" class="section level4">
<h4 class="hasAnchor">
<a href="#coverage-by-query-sequence" class="anchor"></a>Coverage by query sequence</h4>
<p>Sometimes, in addition to knowing how much of a given genome is represented in a whole genome alignment, it is interesting to know which sequences from the other genome are aligning. For example, we might want to ‘paint’ regions of the target sequence by which query seqeunce they form an alignment to. The <code>plot_coverage</code> function can shade the alignment blocks it produces according to any column in the alignment. Here is how we do it for query sequence name (i.e. the <code>qname</code> column), taking advantage of <code>ggplot</code>’s <code>scale_colour_brewer</code> to use a nice palete.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_coverage.html">plot_coverage</a></span>(prim_alignment, <span class="dt">fill=</span><span class="st">'qname'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">palette=</span><span class="st">"Set1"</span>)</code></pre></div>
<p><img src="Introduction_to_pafr_files/figure-html/cov_colourful-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#reading-data-with-read_paf">Reading data with <code>read_paf</code></a></li>
      <li><a href="#filtering-and-subsetting-alignments">Filtering and subsetting alignments</a></li>
      <li><a href="#plotting-alignments">Plotting alignments</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Winter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
